designer/client/src/components/toolbars/scenarioDetails/ScenarioLabels.tsx

```diff
@@ -0,0 +1,315 @@
+};
+
+function toLabelOption(value: string): LabelOption {
+    return {
+        title: value,
+        value: value,
+    };
+}
+
+function toLabelValue(option: LabelOption): string {
+    return option.value;
+}
+
+function formatErrors(errors: ScenarioLabelValidationError[]): string {
+    return errors.map((error) => `Incorrect value '${error.label}': ${error.messages.join(",")}`).join("\n");
+}
+
+export const ScenarioLabels: LabelsEdit = ({ readOnly }: Props) => {
+    const scenarioLabels: LabelOption[] = useSelector(getScenarioLabels).map(toLabelOption);
+    const initialLabelErrors = useSelector(getScenarioLabelsErrors).filter((error) =>
+        scenarioLabels.some((label) => toLabelValue(label) === error.label),
+    );
+    const [labelsErrors, setLabelsErrors] = useState<ScenarioLabelValidationError[]>(initialLabelErrors);
+    const [showEditor, setShowEditor] = useState(scenarioLabels.length !== 0);
+
+    const theme = useTheme();
+    const { menuOption } = selectStyled(theme);
+    const dispatch = useDispatch();
+
+    const [isFetching, setIsFetching] = useState(false);
+    const [isOpen, setIsOpen] = useState(false);
+    const [isEdited, setIsEdited] = useState(false);
+    const [options, setOptions] = useState<LabelOption[]>([]);
+
+    const [input, setInput] = useState("");
+    const [inputTyping, setInputTyping] = useState(false);
+    const [inputErrors, setInputErrors] = useState<ScenarioLabelValidationError[]>([]);
+
+    const handleAddLabelClick = () => {
+        setShowEditor(true);
+        setIsEdited(true);
+    };
+
+    const inputHelperText: () => string = () => {
+        if (inputErrors.length !== 0) {
+            return formatErrors(inputErrors.concat(labelsErrors));
+        }
+
+        if (!isOpen && labelsErrors.length !== 0) {
+            return formatErrors(labelsErrors);
+        }
+
+        return undefined;
+    };
+
+    const validateInput = useMemo(() => {
+        return debounce(async (newInput: string) => {
+            if (newInput !== "") {
+                const response = await HttpService.validateScenarioLabels([newInput]);
+
+                if (response.status === 200) {
+                    setInputErrors(response.data.validationErrors);
+                }
+            }
+
+            setInputTyping(false);
+        }, 500);
+    }, []);
+
+    const validateLabels = useMemo(() => {
+        return debounce(async (labels: LabelOption[]) => {
+            const values = labels.map(toLabelValue);
+            const response = await HttpService.validateScenarioLabels(values);
+
+            if (response.status === 200) {
+                const validationError = response.data.validationErrors;
+                setLabelsErrors(validationError);
+            }
+        }, 500);
+    }, []);
+
+    const fetchAvailableLabels = useCallback(async () => {
+        setIsFetching(true);
+        const { data } = await HttpService.fetchScenarioLabels();
+        setIsFetching(false);
```

Comment:
In this approach, there can be a situation, where this flag won't be set to true, for example, if fetchScenarioLabels throws an error. Can you use try {} finally {} here and move setIsFetching(false); to the finally block?